docs/design/cloud-init.rst

```diff
@@ -0,0 +1,353 @@
+  fit into the particular environment it is being run in [#f2]_.
+
+Unfortunately, older versions of cloud-init don't play well with
+``vendor-data``, which leaves us with ``meta-data`` and the files under
+``/etc/cloud/cloud.cfg.d/``.
+
+Configuration Tasks
+-------------------
+
+*snf-image* configures the VM by running a number of configuration tasks on the
+hard disk of the VM. On images that support cloud-init, instead of directly
+altering the needed system files, *snf-image-helper* could add cloud-init
+configuration into the ``meta-data`` file as well as files under
+``/etc/cloud/cloud.cfg.d/``.
+
+The proposed changes for the configuration tasks are:
+
+**10FixPartitionTable**: The task should be prevented from running. Cloud-init
+will automatically grow the partition to consume the available space using the
+``growroot`` module of cloud-initramfs-tools and ``cc_growpart`` module.
+
+**20FilesystemResizeUnmounted**: This task should be prevented from running.
+Cloud-init will automatically enlarge the file system using the ``cc_resizefs``
+module.
+ 
+**30MountImage**: No changes are needed for this task.
+
+**35InstallUnattend**: This can stay intact or be prevented from running. It's
+Windows specific task and won't do anything against a Linux image.
+
+**40FilesystemResizeMounted**: This should be prevented from running.
+Cloud-init will use ``cc_resizefs`` module to grow the file system.
+
+**50AddSwap**: This task parses the content of the *SWAP* image property which
+cames in two forms (``<partition id>:<size>`` or ``<disk letter>``) and either
+creates a new swap partition or formats a whole disk to be used as swap. In 
+cloud-init, the swap configuration is performed by the ``cc_mount`` module.
+This module has a swap config key that can be used like this:
+
+.. code-block:: yaml
+
+  swap:
+    filename: <file>
+    size: <"auto"/size in bytes>
+    maxsize: <size in bytes>
+
+With this key we can add a swap file but not a swap partition. Since cloud-init
+does the partitioning itself and using swap files is as efficient as using
+partitions, in case the *SWAP* image property is defined in the first form,
+it's better if we just ignore the partition id and use the swap config key to
+create a swap file of the requested size instead.
+
+In case the *SWAP* property is defined in the second form and a swap disk is
+requested, we could make use of the ``mounts`` key of the ``cc_mounts`` module,
+to put the appropriate entry in fstab:
+
+
+.. code-block:: yaml
+
+  mounts:
+    - [ /dev/ephemeral-1, /mnt, auto, "defaults,noexec" ]
+    - [ sdc, /opt/data ]
+    - [ xvdh, /opt/data, "auto", "defaults,nofail", "0", "0" ]
+
+Unfortunately, this module does not support providing non ephemeral device
+names (UUID for swap) and using the standard device naming is error-prone.
+Hence, for swap disks, its better if we bypass cloud-init and let snf-image
+directly format the disk and put the swap entry to fstab.
```

Comment:
I am not really fond of using cloud-init for swap files and an snf-image task for swap disks. It makes the code/logic a bit more complicated I think. 
Since snf-image/cloud-init is just for provisioning the image, and device names won't change during provisioning, why do you want to avoid using cloud-init with static device names for the swap disk ? I'd be happier if cloud-init supported UUIDs but I can't see a problem not using them for our usecase.